### PR TITLE
Simple arithmetic

### PR DESCRIPTION
**Pull request by derekpappas**
### Checklist

- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?

### Type of Improvement
Simple add and subtract operations with negative literals are inverted.

- new feature/API

### API Impact
None

See above

### Backend Code Generation Impact
   ```
     circuit Top :
     module Top :
       input a : SInt<2>
       output y : SInt<2>
       y <= add(a, SInt(-1))
   ```
  
   converts to
  
   ```
     circuit Top :
     module Top :
       input a : SInt<2>
       output y : SInt<2>
       y <= sub(a, SInt(1))
   ```
  
   ________________________
  
    ```
        circuit Top :
        module Top :
          input a : SInt<2>
          output y : SInt<2>
          y <= sub(a, SInt(-1))
      ```
   
    converts to
   
    ```
        circuit Top :
        module Top :
          input a : SInt<2>
          output y : SInt<2>
          y <= add(a, SInt(1))
      ```

See above

### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference.

### CODE REVIEW

**Comment by aswaterman** [comment]
Negating -2^(W-1) will overflow a W-bit SInt.  Do you need to prevent this conversion in that case?

**Comment by derekpappas** [comment]
> Negating -2^(W-1) will overflow a W-bit SInt. Do you need to prevent this conversion in that case?

yes. my bad.

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    e match {
+      case DoPrim(Add, Seq(e, lit@SIntLiteral(n, _)), consts: Seq[BigInt], tpe) if (n < 0) =>  DoPrim(Sub, Seq(e, lit.copy(n * -1)), consts, tpe)
+      case DoPrim(Addw, Seq(e, lit@SIntLiteral(n, _)), consts: Seq[BigInt], tpe) if (n < 0) =>  DoPrim(Subw, Seq(e, lit.copy(n * -1)), consts, tpe)
+      case DoPrim(Sub, Seq(e, lit@SIntLiteral(n, IntWidth(w))), consts: Seq[BigInt], tpe) if (n < 0) && notMax(n, w) =>
+        DoPrim(Add, Seq(e, lit.copy(n * -1)), consts, tpe)
+      case DoPrim(Subw, Seq(e, lit@SIntLiteral(n, IntWidth(w))), consts: Seq[BigInt], tpe) if (n < 0) && notMax(n, w) =>
+        DoPrim(Addw, Seq(e, lit.copy(n * -1)), consts, tpe)
+      case _ => e
+    }
+  }
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: When matching on `consts`, you don't need to specify the type `Seq[BigInt]` because it's always that type.

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  *   *     y <= add(a, SInt(1))
+  *   * }}}
+   */
+class SimpleArithmetic extends Transform {
+  def inputForm = LowForm
+  def outputForm = LowForm
+
+  import SimpleArithmetic._
+
+  def makeModuleInfo(c: Circuit): (Map[String, Map[String, String]], Seq[String], InstanceGraph, Map[String, DefModule]) = {
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: I don't think you call this function - copy pasta?

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        case DefNode(_, nname, rhs) => netlist(WrappedExpression(WRef(nname))) = rhs
+        case _ => // Do nothing
+      }
+      stmt
+    }
+    mod.map(onStmt)
+    netlist
+  }
+
+  def buildSymTab(mod: DefModule): Map[WrappedExpression, Int] = {
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: Perhaps we should wait until after we merge the other PR, so you can just reuse that function instead of creating a new, identical one?

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    val modNameModules = x.map{ m => (m.name -> m)}.toMap
+    (childInstances, moduleNameTraversalOrder, iGraph, modNameModules)
+  }
+
+  /** Build a [[Netlist]] from a Module's connections and Nodes
+    * This assumes [[firrtl.LowForm LowForm]]
+    * @param mod [[firrtl.ir.Module Module]] from which to build a [[Netlist]]
+    * @return [[Netlist]] of the module's connections and nodes
+    */
+  def buildNetlist(mod: DefModule): Netlist = {
```
Comment (azidar) [inline src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: Perhaps we should wait until after we merge your other PR, so that we can reuse this function?

File: src/test/scala/firrtlTests/SimpleArtihmeticTests.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
+// See LICENSE for license details.
+
+package firrtlTests
+
+import firrtl.transforms.SimpleArithmetic
+
+/**
+ * Tests inline instances transformation
+ */
+class SimpleArtihmeticTests extends LowTransformSpec {
```
Comment (azidar) [inline src/test/scala/firrtlTests/SimpleArtihmeticTests.scala]: Add test case for overflow (max integer).

File: src/test/scala/firrtlTests/SimpleArtihmeticTests.scala [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      """circuit Top :
+        |  module Top :
+        |    input a : SInt<2>
+        |    output y : SInt<2>
+         |    y <= asSInt(bits(add(a, SInt<1>("h1")), 1, 0))
+        """.stripMargin
+    execute(input, check, Nil)
+  }
+
+}
```
Comment (azidar) [inline src/test/scala/firrtlTests/SimpleArtihmeticTests.scala]: Add newline :)

File: src/test/scala/firrtlTests/SimpleArtihmeticTests.scala (hunk 5 above)
Comment (derekpappas) [reply src/test/scala/firrtlTests/SimpleArtihmeticTests.scala]: @azidar @aswaterman Check that my test that I added is correct. For a 2 bit signal I put -3.

```
  "The module Top" should "not convert sub(_, -<max value>) to add(_, 1)" in {
    val input =
      """circuit Top :
        |  module Top :
        |    input a : SInt<2>
        |    output y : SInt<2>
        |    y <= sub(a, SInt(-3))""".stripMargin
    val check =
      """circuit Top :
        |  module Top :
        |    input a : SInt<2>
        |    output y : SInt<2>
        |    y <= asSInt(bits(sub(a, SInt(-3))), 1, 0))
        """.stripMargin
    execute(input, check, Nil)
  }
```

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+  /**
+    *
+    * @param e
+    * @return
+    */
+  def replaceSimple(e: Expression): Expression = {
+    e match {
+      case DoPrim(Add, Seq(e, lit@SIntLiteral(n, _)), consts: Seq[BigInt], tpe) if (n < 0) =>  DoPrim(Sub, Seq(e, lit.copy(n * -1)), consts, tpe)
+      case DoPrim(Addw, Seq(e, lit@SIntLiteral(n, _)), consts: Seq[BigInt], tpe) if (n < 0) =>  DoPrim(Subw, Seq(e, lit.copy(n * -1)), consts, tpe)
```
Comment (aswaterman) [inline src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: You need the `notMax` check for these two, as well.

File: src/test/scala/firrtlTests/SimpleArtihmeticTests.scala (hunk 5 above)
Comment (aswaterman) [reply src/test/scala/firrtlTests/SimpleArtihmeticTests.scala]: You want `SInt(-2)` here.

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala (hunk 3 above)
Comment (derekpappas) [reply src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: In the CreateClockEnables transform PR I have a new file called IRUtils.scala where I am putting all sorts of utility functions. The problem is that there are 4? firrtl PR's in flight and some are in free chips and some are in sifive. We need to decide how to manage these utility functions.

File: src/main/scala/firrtl/transforms/SimpleArithmetic.scala (hunk 2 above)
Comment (derekpappas) [reply src/main/scala/firrtl/transforms/SimpleArithmetic.scala]: Yes

File: src/main/scala/firrtl/Utils.scala [hunk 8]
```diff
@@ -210,6 +210,17 @@ object Utils extends LazyLogging {
...
+  /** Returns true if PrimOp is a isSimple Arithmetic Operator, false otherwise */
+  def isSimpleArithmeticExpr(op: PrimOp): Boolean = op match {
+    case Add | Sub  => true
+    case _ => false
+  }
+  /** Returns true if Expression is a Simple Arithmetic PrimOp, false otherwise */
+  def isSimpleArithmeticExpr (expr: Expression): Boolean = expr match {
+    case DoPrim(op, _,_, UIntType(_)) if isSimpleArithmeticExpr(op) => true
+    case _ => false
+  }
```
[the remark this answers is not in the document]
Comment (derekpappas) [reply src/main/scala/firrtl/Utils.scala]:
```
 def isAddorSubExpr (expr: Expression): Boolean = {
    val b = expr match {
      case DoPrim(Add, _,_, UIntType(_)) => true
      case DoPrim(Sub, _,_, UIntType(_)) => true
      case DoPrim(Add, _,_, SIntType(_)) => true
      case DoPrim(Sub, _,_, SIntType(_)) => true
      case _ => false
    }
    b
  }
```

**Comment by azidar** [comment]
Subsumed by #1374
